File: coding_java/app/Logger.java
/**
 * This class provides log wrappers for the classic system signals.
 */
public class Logger {

  public static void logSuccess() {
    System.out.println("----------------------------------------------------");
    System.out.println("Everything went well!");
    System.out.println("----------------------------------------------------");
  }
  
  public static void logNoMatch(String... args) {
    System.err.println("----------------------------------------------------");
    System.err.println("The result does not match the expected value");
    System.err.println("Expected: " + args[0]);
    System.err.println("Actual: " + args[1]);
    System.err.println("----------------------------------------------------");
  }
  
  public static void logException(String... args) {
    System.err.println("----------------------------------------------------");
    System.err.println("An error occured during runtime.");
    System.err.println("This is all your fault. Shame. SHAME!");
    System.err.println("Details:");
    System.err.println("Expected: " + args[0]);
  }
  
  public static void logException(Throwable throwable) {
    System.err.println("----------------------------------------------------");
    System.err.println("Something bad happened!");
    System.err.println(throwable.getMessage());
    throwable.printStackTrace();
    System.err.println("----------------------------------------------------");
  }
  
  
}

File: coding_cpp/app/logger.cpp
/**
 * This class provides log wrappers for the classic system signals.
 */
#include "logger.h"
#include <iostream>
#include <string>

using std::string;
using std::endl;
using std::cout;
using std::cerr;
using std::exception;

void Logger::log_success() {
  cout << "----------------------------------------------------" << endl;
  cout << "Everything went well!" << endl;
  cout << "----------------------------------------------------" << endl;
}

void Logger::log_no_match(const string& expected, const string& actual) {
  cerr << "----------------------------------------------------" << endl;
  cerr << "The result does not match the expected value" << endl;
  cerr << "Expected: " << expected << endl;
  cerr << "Actual: " << actual << endl;
  cerr << "----------------------------------------------------" << endl;
}

void Logger::log_unexpected_error(const string& expected) {
  cerr << "----------------------------------------------------" << endl;
  cerr << "An error occured at runtime." << endl;
  cerr << "This is all your fault. Shame. SHAME!" << endl;
  cerr << "Details:" << endl;
  cerr << "Expected: " << expected << endl;
  cerr << "----------------------------------------------------" << endl;
}

void Logger::log_exception(const exception& e) {
  cerr << "----------------------------------------------------" << endl;
  cerr << "Something bad happened." << endl;
  cerr << "Details: " << e.what() << endl;
  cerr << "----------------------------------------------------" << endl;
}
